### PR TITLE
mgr/dashboard: add tags to edit bucket

### PR DESCRIPTION
**Pull request by Pegonzal**
Add the tags field when creating and editing a bucket

image

Fixes: https://tracker.ceph.com/issues/63412
Signed-off-by: Pedro Gonzalez Gomez <pegonzal@redhat.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by cloudbehl** [comment]
Hi @Pegonzal, Thanks for PR.

As tags are key value pairs, we can represent them in same manner. A text box for key and a text box for value and with a plus sign if a user wants to add more.

[   Key    ]    = [   Value   ]   +(to add more key & values)

image


Similar to how we did this with ceph user adding  capabilities.

**Comment by Pegonzal** [comment]
> Hi @Pegonzal, Thanks for PR.
> 
> As tags are key value pairs, we can represent them in same manner. A text box for key and a text box for value and with a plus sign if a user wants to add more.
> 
> [ Key ] = [ Value ] +(to add more key & values)
> 
> image
> 
> Similar to how we did this with ceph user adding capabilities.

Updated it!

File: src/pybind/mgr/dashboard/services/rgw_client.py [hunk 1]
```diff
@@ -702,6 +702,16 @@ def set_bucket_encryption(self, bucket_name, key_id,
         except RequestException as e:
             raise DashboardException(msg=str(e), component='rgw')
 
+    @RestClient.api_put('/{bucket_name}?tagging')
+    def set_tags(self, bucket_name, tags, request=None):
+        # pylint: disable=unused-argument
+        try:
+            ET.fromstring(tags)
+        except ET.ParseError:
+            return "Data must be properly formatted"
```
Comment (avanthakkar) [inline src/pybind/mgr/dashboard/services/rgw_client.py:711]: Does it makes sense to give hint for what format we're expecting?

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.html [hunk 2]
```diff
@@ -385,6 +385,33 @@
           </div>
         </fieldset>
 
+        <!-- Tags -->
+        <fieldset *ngIf="editing">
+          <legend class="cd-header"
+                  i18n>Tags
```
Comment (avanthakkar) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.html:391]: A helper would be nice to know what tags represents here,

Comment (Pegonzal) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.html:391]: Sure, I'll add it to explain the Tags function

Comment (Pegonzal) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.html:391]: Added a helper to provide some context

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.html [hunk 3]
```diff
@@ -395,3 +422,38 @@
...
+    <button type="button"
+            class="btn btn-light"
+            id="tag-delete-{{index}}"
+            i18n-ngbTooltip
+            ngbTooltip="Delete"
+            (click)="deleteTag(index)">
+      <i [ngClass]="[icons.trash]"></i>
+    </button>
+  </div>
+  <span class="help-block"></span>
```
[the remark this answers is not in the document]
Comment (Pegonzal) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.html]: Remmoved it as it was unnecessary

File: src/pybind/mgr/dashboard/controllers/rgw.py [hunk 4]
```diff
@@ -279,6 +279,9 @@ def _set_locking(self, owner, daemon_name, bucket_name, mode,
     def _get_policy(self, bucket: str):
         rgw_client = RgwClient.admin_instance()
         return rgw_client.get_bucket_policy(bucket)
+    def _set_tags(self, bucket_name, tags, daemon_name, owner):
```
[the remark this answers is not in the document]
Comment (Pegonzal) [reply src/pybind/mgr/dashboard/controllers/rgw.py:283]: Yes it will, fixed it, thanks

**Comment by cloudbehl** [comment]
Thanks for update on PR @Pegonzal, Looks and works really good.

Two things:
1. When we click on add tags the title of modal says `Set tags` and I think it should be `Add Tags` only. 
image
2. Details page should have visualization of tags Which are set.

**Comment by Pegonzal** [comment]
> Thanks for update on PR @Pegonzal, Looks and works really good.
> 
> Two things:
> 
> 1. When we click on add tags the title of modal says `Set tag` and I think it should be `Add Tag` only.
>    image
Now it shows Add Tag or Edit Tag when adding a tag or editing one respectively
> 2. Details page should have visualization of tags Which are set.
Added the tags to the details page

**Review by avanthakkar — APPROVED** [review]
LGTM!

**Comment by cloudbehl** [comment]
Thanks @Pegonzal, the PR looks really good. 

Only one thing, we can have tags a part of create bucket as well.

**Review by cloudbehl — APPROVED** [review]
LGTM

**Comment by Pegonzal** [comment]
Added some better form validators, the PR should be good to go know

**Pull request merged**